crates/starlight/src/gc/snapshot.rs

```diff
@@ -36,9 +36,32 @@ impl Snapshot {
         let buffer = serializer.output.into_boxed_slice();
         serializer.output = vec![];
 
-        Snapshot {
-            buffer,
-            serializer
-        }
+        Snapshot { buffer, serializer }
+    }
+
+    pub fn take_context(
+        log: bool,
+        runtime: &mut Runtime,
+        context: &mut Context,
+        callback: impl FnOnce(&mut SnapshotSerializer, &mut Context),
+    ) -> Self {
+        let mut serializer = serializer::SnapshotSerializer::new(log);
+        let ids_patch = serializer.output.len();
+        serializer.write_u32(0);
+        serializer.build_reference_map(runtime);
+        serializer.build_symbol_table();
+        serializer.build_heap_reference_map_in_context(runtime, context);
```

Comment:
single build_reference_map call is enough. You don't have to walk it a second time in context. Also building a reference map from context will not work, you have to iterate entire heap always.

Reply:
Actually, now build_heap_reference_map_in_context will only trace objects in the specific context. And build_reference_map will only record native references?

Reply:
Ah right, I forgot that there is another function for this, sorry